src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,345 @@
+
+	private final ConsumerFactory<K, V> consumerFactory;
+
+	private final KafkaAckCallbackFactory<K, V> ackCallbackFactory;
+
+	private final String[] topics;
+
+	private String groupId;
+
+	private String clientId = "message.source";
+
+	private long pollTimeout = DEFAULT_POLL_TIMEOUT;
+
+	private RecordMessageConverter messageConverter = new MessagingMessageConverter();
+
+	private Type payloadType;
+
+	private volatile Consumer<K, V> consumer;
+
+	private volatile Iterator<ConsumerRecord<K, V>> recordIterator;
+
+	private volatile Collection<TopicPartition> partitions;
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory, String... topics) {
+		this(consumerFactory, new KafkaAckCallbackFactory<K, V>(), topics);
+	}
+
+	public KafkaMessageSource(ConsumerFactory<K, V> consumerFactory,
+			KafkaAckCallbackFactory<K, V> ackCallbackFactory, String... topics) {
+		Assert.notNull(consumerFactory, "'consumerFactory' must not be null");
+		Assert.notNull(ackCallbackFactory, "'ackCallbackFactory' must not be null");
+		this.consumerFactory = consumerFactory;
+		this.ackCallbackFactory = ackCallbackFactory;
+		this.topics = topics;
+	}
+
+	protected String getGroupId() {
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the messsage converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized AbstractIntegrationMessageBuilder<Object> doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		if (this.recordIterator != null && this.recordIterator.hasNext()) {
+			record = this.recordIterator.next();
```

Comment:
Does it really work as expected when subsequent `doReceive()` comes from a different threads than the initial ` this.consumer.poll()`?
I mean don't we need to exhaust the iterator to the internal `Queue` otherwise? And therefore we won't need pause and reset 😄

Reply:
Yes; we can use the iterator; it's all synchronized - no need to copy to a queue.

Reply:
OTOH maybe we should enforce `max.poll.records = 1` to avoid any pause/resume and this iterator stuff altogether?

Reply:
That was my initial implementation but I decided it wasn't a good idea to require a round trip to the broker for each message.

Reply:
OK. But you do pause/resume otherwise...

Reply:
? We have to since we have no control over the caller poll frequency.

Reply:
>OTOH maybe we should enforce max.poll.records = 1 to avoid any pause/resume and this iterator stuff altogether?

I've decided to go back to that model - there are several issues with resetting/seeking otherwise.

Reply:
OK. Since I don't see changed, I assume that was just an information and the fix is coming this week.

Reply:
I am still working on this; almost done (might be tomorrow now), but it's a major change, so I think squashing makes most sense. Let me know if you prefer to keep history.

Reply:
Squash sounds good, but really doesn't matter for me.
I worry more about commit comment to understand what's going on.
Thanks